test/src/protocol/RemotingCommandTest.cpp

```diff
@@ -0,0 +1,285 @@
+}
+
+TEST(remotingCommand, encodeAndDecode) {
+    RemotingCommand remotingCommand(13, "CPP", MQVersion::s_CurrentVersion, 12, 3, "remark", NULL);
+    remotingCommand.SetBody("123123", 6);
+    remotingCommand.Encode();
+    // no delete
+    const MemoryBlock *head = remotingCommand.GetHead();
+    const MemoryBlock *body = remotingCommand.GetBody();
+
+    unique_ptr<MemoryOutputStream> result(new MemoryOutputStream(1024));
+    result->write(head->getData() + 4, head->getSize() - 4);
+    result->write(body->getData(), body->getSize());
+
+    shared_ptr<RemotingCommand> decodeRemtingCommand(RemotingCommand::Decode(result->getMemoryBlock()));
+    EXPECT_EQ(remotingCommand.getCode(), decodeRemtingCommand->getCode());
+    EXPECT_EQ(remotingCommand.getOpaque(), decodeRemtingCommand->getOpaque());
+    EXPECT_EQ(remotingCommand.getRemark(), decodeRemtingCommand->getRemark());
+    EXPECT_EQ(remotingCommand.getVersion(), decodeRemtingCommand->getVersion());
+    EXPECT_EQ(remotingCommand.getFlag(), decodeRemtingCommand->getFlag());
+    EXPECT_TRUE(decodeRemtingCommand->getCommandHeader() == NULL);
+
+    // ~RemotingCommand delete
+    GetConsumerRunningInfoRequestHeader *requestHeader = new GetConsumerRunningInfoRequestHeader();
+    requestHeader->setClientId("client");
+    requestHeader->setConsumerGroup("consumerGroup");
+    requestHeader->setJstackEnable(false);
+
+    RemotingCommand encodeRemotingCommand(307, "CPP", MQVersion::s_CurrentVersion, 12, 3, "remark", requestHeader);
+    encodeRemotingCommand.SetBody("123123", 6);
+    encodeRemotingCommand.Encode();
+
+    // no delete
+    const MemoryBlock *phead = encodeRemotingCommand.GetHead();
+    const MemoryBlock *pbody = encodeRemotingCommand.GetBody();
+
+    unique_ptr<MemoryOutputStream> results(new MemoryOutputStream(1024));
+    results->write(phead->getData() + 4, phead->getSize() - 4);
+    results->write(pbody->getData(), pbody->getSize());
+
+    shared_ptr<RemotingCommand> decodeRemtingCommandTwo(RemotingCommand::Decode(results->getMemoryBlock()));
+
+    decodeRemtingCommandTwo->SetExtHeader(encodeRemotingCommand.getCode());
+    GetConsumerRunningInfoRequestHeader *header =
+        reinterpret_cast<GetConsumerRunningInfoRequestHeader *>(decodeRemtingCommandTwo->getCommandHeader());
+    EXPECT_EQ(requestHeader->getClientId(), header->getClientId());
+    EXPECT_EQ(requestHeader->getConsumerGroup(), header->getConsumerGroup());
+}
+
+TEST(remotingCommand, SetExtHeader) {
+    shared_ptr<RemotingCommand> remotingCommand(new RemotingCommand());
+
+    remotingCommand->SetExtHeader(-1);
+    EXPECT_TRUE(remotingCommand->getCommandHeader() == NULL);
+
+    Json::Value object;
+    Json::Value extFields;
+    extFields["id"] = 1;
+    object["extFields"] = extFields;
+    remotingCommand->setParsedJson(object);
+
+    remotingCommand->SetExtHeader(MQRequestCode::SEND_MESSAGE);
+    SendMessageResponseHeader *sendMessageResponseHeader =
```

Comment:
sendMessageResponseHeader is NULL? also the next  ptrs, so this Test will crash?

Reply:
sendMessageResponseHeader  is not null , this is formatted result

Reply:
what's the mean of  formatted result, i think it will nullptr, because the is no place to set value to m_pExtHeader 

RemotingCommand() : m_code(0) {}; -->m_pExtHeader is nullptr or undefined
remotingCommand->getCommandHeader() -- > return m_pExtHeader

Reply:
here no problem， because SetExtHeader will set the m_pExtHeader, so it will not null for remotingCommand->getCommandHeader()